### PR TITLE
Core : Catalog Tables Migration API

### PR DESCRIPTION
**Pull request by Mehul2500**
Introducing `migrateTables()` in `CatalogUtil` which could help in the migration of Iceberg tables in any Source Catalog to any Target Catalog. Uses PR #5037 , for the `registerTable()` functionality  in `BaseMetastoreCatalog`.

I used tables migrating from Hadoop Catalog to Hive Catalog for the test case.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 1]
```diff
@@ -346,4 +353,63 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
+    Catalog targetCatalog;
+    try {
+      targetCatalog = loadCatalog(targetCatalogProperties.get("catalogImpl"),
+          targetCatalogProperties.get("catalogName"), targetCatalogProperties, targetHadoopConfig);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot initialize Target Catalog implementation %s: %s", targetCatalogProperties.get("catalogImpl"),
+          e.getMessage()), e);
+    }
+    List<TableIdentifier> allIdentifiers = tableIdentifiers;
```
Comment (snazy) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:394]: I think, this code should probably live in `Catalog`:  A new function like `Catalog.registerTableFromCatalog()` to "move" a single table to the current catalog. The `HadoopCatalog` could then do the special-handling in its implementation.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 2]
```diff
@@ -346,4 +353,63 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
+    if (tableIdentifiers == null || tableIdentifiers.isEmpty()) {
+      List<Namespace> namespaces = (sourceCatalog instanceof SupportsNamespaces) ?
+          ((SupportsNamespaces) sourceCatalog).listNamespaces() : ImmutableList.of(Namespace.empty());
+      allIdentifiers = namespaces.stream().flatMap(ns ->
+          sourceCatalog.listTables(ns).stream()).collect(Collectors.toList());
+    }
+    List<TableIdentifier> migratedTableIdentifiers = new ArrayList<TableIdentifier>();
+    allIdentifiers.forEach(tableIdentifier -> {
```
Comment (snazy) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:402]: I suspect this will run for a very long time, like when there a a lot of tables.
If things fail in the meantime, it's hard to resume after the failed table.
I.e. error handling here is tricky.

Not sure whether it is actually possible to properly handle the case when `registerTable` worked, but `dropTable` failed - in such a case you'd have the same table in two catalogs.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 3]
```diff
@@ -346,4 +353,63 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot initialize Target Catalog implementation %s: %s", targetCatalogProperties.get("catalogImpl"),
+          e.getMessage()), e);
+    }
+    List<TableIdentifier> allIdentifiers = tableIdentifiers;
+    if (tableIdentifiers == null || tableIdentifiers.isEmpty()) {
+      List<Namespace> namespaces = (sourceCatalog instanceof SupportsNamespaces) ?
+          ((SupportsNamespaces) sourceCatalog).listNamespaces() : ImmutableList.of(Namespace.empty());
+      allIdentifiers = namespaces.stream().flatMap(ns ->
```
Comment (snazy) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:398]: I suspect this will run for a very long time, like when there a a lot of tables.

File: core/src/main/java/org/apache/iceberg/CatalogUtil.java [hunk 4]
```diff
@@ -346,4 +353,63 @@ public static void configureHadoopConf(Object maybeConfigurable, Object conf) {
...
+   * @param sourceCatalogProperties Source Catalog Properties
+   * @param targetCatalogProperties Target Catalog Properties
+   * @param sourceHadoopConfig      Source Catalog Hadoop Configuration
+   * @param targetHadoopConfig      Target Catalog Hadoop Configuration
+   * @return list of table identifiers for successfully migrated tables
+   */
+  public static List<TableIdentifier> migrateTables(List<TableIdentifier> tableIdentifiers,
+      Map<String, String> sourceCatalogProperties, Map<String, String> targetCatalogProperties,
+      Object sourceHadoopConfig, Object targetHadoopConfig) {
+    if (tableIdentifiers != null) {
```
Comment (snazy) [inline core/src/main/java/org/apache/iceberg/CatalogUtil.java:372]: Let's leave out the catalog instantiation and configuration here completely. I suspect that users have at least one of these catalogs already handy - and setting up "the same" catalog twice is superfluous.

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java [hunk 5]
```diff
@@ -152,7 +152,8 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       }
     }
 
-    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+    String newMetadataLocation = (base == null) && (metadata.metadataFileLocation() != null) ?
```
Comment (snazy) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:155]: Why is this necessary?
It's a new commit, not sure whether it is good that it re-uses an existing metadata location that is (potentially) "owned" by another catalog.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 6]
```diff
@@ -385,6 +388,89 @@ public void testDropTable() throws IOException {
     verifyCommitMetadata();
   }
 
+  private void validateRegister(TableIdentifier identifier, String metadataVersionFiles) {
+    Assertions.assertThat(catalog.registerTable(identifier, "file:" + metadataVersionFiles)).isNotNull();
+    Table newTable = catalog.loadTable(identifier);
+    Assertions.assertThat(newTable).isNotNull();
+    TableOperations ops = ((HasTableOperations) newTable).operations();
+    String metadataLocation = ((NessieTableOperations) ops).currentMetadataLocation();
+    Assertions.assertThat("file:" + metadataVersionFiles).isEqualTo(metadataLocation);
```
Comment (snazy) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:397]: Hint: Your assertions are often the "wrong" way around.
It's always `assertThat(<current state>)....`, followed by the expectations.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 7]
```diff
@@ -385,6 +388,89 @@ public void testDropTable() throws IOException {
...
+    Assertions.assertThat(catalog.dropTable(identifier, false)).isTrue();
+  }
+
+  @Test
+  public void testRegisterTableWithGivenBranch() {
+    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    Assertions.assertThat(1).isEqualTo(metadataVersionFiles.size());
```
Comment (snazy) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:404]: Hint: `assertThat(metadataVersionFiles).hasSize(1)`

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 8]
```diff
@@ -385,6 +388,89 @@ public void testDropTable() throws IOException {
...
+    ImmutableTableReference tableReference =
+        ImmutableTableReference.builder().reference("main").name(TABLE_NAME).build();
```
Comment (snazy) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:406]: Please use a different branch here.
Using the default branch is not that great - and the test says `...Branch` implying it's not the default branch.

File: core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java [hunk 9]
```diff
@@ -639,4 +642,29 @@ public void testConversions() {
     Assert.assertEquals(ns, JdbcUtil.stringToNamespace(nsString));
   }
 
+  @Test
```
Comment (snazy) [inline core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java:645]: Should these tests better live in `CatalogTests`?

File: aws/src/integration/java/org/apache/iceberg/aws/dynamodb/TestDynamoDbCatalog.java [hunk 10]
```diff
@@ -295,6 +298,37 @@ public void testDropNamespace() {
     Assert.assertFalse("namespace must not exist", response.hasItem());
   }
 
+  @Test
+  public void testRegisterTable() {
```
Comment (snazy) [inline aws/src/integration/java/org/apache/iceberg/aws/dynamodb/TestDynamoDbCatalog.java:302]: This pair of tests is repeated (in a very similar way) across multiple catalogs. Can those be centralized somewhere? `CatalogTests` maybe?

**Comment by ajantha-bhat** [comment]
@Mehul2500 : Please rebase this PR.

**Comment by ajantha-bhat** [comment]
As there is no activity in this. 
I have opened https://github.com/apache/iceberg/pull/5492
